<a href="https://colab.research.google.com/github/Decoding-Data-Science/nov25/blob/main/18th_Nov_of_logistic_regression1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Python Code: Age, Income, and Purchase Prediction
# 1. SETUP: Import necessary libraries
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.datasets import make_classification # Tool for generating synthetic data

print("Step 1: Setting up Logistic Regression for the Purchase Prediction.")


Step 1: Setting up Logistic Regression for the Purchase Prediction.


In [ ]:
X, y = make_classification(
    n_samples=1000,
    n_features=2,           # Represents two inputs: Age and Income [3]
    n_informative=2,        # Both features influence the outcome
    n_redundant=0,
    n_classes=2,            # Binary classification (Purchase/No Purchase) [2]
    random_state=42
)

# Rename features for better interpretability later
feature_names = ['Age (Feature 1)', 'Income (Feature 2)']

print(f"Sample Input (Age, Income): {X[:1]}")
print(f"Sample Answer (Purchase): {y[:1]}")

# 3. DATA ORGANIZATION: Splitting for Training and Testing
# We separate the "clues" (X) from the "answers" (Y) [1].
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42
)

# 4. MODEL CREATION AND TRAINING
# We grab the Logistic Regression model from the scikit-learn toolbox [4].
model = LogisticRegression(random_state=42)

# This is the "magic moment" where all the learning happens using the .fit command.
# The model learns the pattern between Age/Income (X) and Purchase (Y) [1].
model.fit(X_train, y_train)

print("\nStep 4: Model training complete.")


# 5. TESTING/PREDICTION
# We use the predict command to ask the model for its best guess (0 or 1) on new data [1].
y_pred = model.predict(X_test)



Sample Input (Age, Income): [[-0.99910178 -0.66386   ]]
Sample Answer (Purchase): [1]

Step 4: Model training complete.


In [ ]:

# 6. EVALUATION
# We check how often the model's guess matches the actual answer (y_test).
accuracy = accuracy_score(y_test, y_pred)

print(f"\nStep 6: Model Evaluation Results")
print(f"Accuracy on Test Data: {accuracy:.4f}")


# 7. INTERPRETABILITY (Model Insights)
# Logistic regression has a "genuine superpower": interpretability. It's a "glass box" [5, 6].
print("\nStep 7: Model Insights (Interpretability):")

# Access the coefficients—the learned "weights" from the weighted sum [3].
# In binary classification, coefficients are in the first row of model.coef_.
coefficients = model.coef_

# Display the coefficients for Age and Income
for name, coef_value in zip(feature_names, coefficients[0]):
    print(f"{name} Coefficient: {coef_value:.4f}")

# The interpretation provides "real insights" [5]
print("\nInterpretation:")
print("A positive coefficient means that feature (Age or Income) increases the probability of a Purchase (1).")
print("A negative coefficient means that feature decreases the probability of a Purchase (1).")
print("This allows us to say in plain English how much each piece of information is tilting the scales [5].")


Step 6: Model Evaluation Results
Accuracy on Test Data: 0.8733

Step 7: Model Insights (Interpretability):
Age (Feature 1) Coefficient: -0.3306
Income (Feature 2) Coefficient: 2.0174

Interpretation:
A positive coefficient means that feature (Age or Income) increases the probability of a Purchase (1).
A negative coefficient means that feature decreases the probability of a Purchase (1).
This allows us to say in plain English how much each piece of information is tilting the scales [5].


1. From coefficient → impact (log-odds vs. probability)

In linear regression:

Coefficient = change in y for 1-unit change in X.

In logistic regression:

Coefficient = change in log-odds of purchase for 1-unit change in X.

So we usually convert them to odds ratios:

odds ratio
=
𝑒
𝛽
odds ratio=e
β
2. Impact of Age (β = -0.3306)

Odds ratio =
𝑒
−
0.3306
≈
0.72
e
−0.3306
≈0.72

Business meaning:

For every +1 unit increase in Age (e.g., +1 year, if Age is in years),
→ the odds of purchase are multiplied by ~0.72
→ i.e., odds drop by about 28%.

So you can say:

“Holding income constant, every additional year of age reduces the odds of purchasing by roughly 28%.”

3. Impact of Income (β = 2.0174)

Odds ratio =
𝑒
2.0174
≈
7.52
e
2.0174
≈7.52

Business meaning:

For every +1 unit increase in Income (whatever unit you used – e.g., 1 normalized unit, or 10k AED if you scaled it that way),
→ the odds of purchase are multiplied by ~7.5
→ i.e., odds increase by around 650%.

So in business language:

“Holding age constant, a one-unit increase in income increases the odds of purchasing by about 7.5×.”

This is the same spirit as linear regression:

Age → negative impact on purchase likelihood.

Income → strong positive impact, and much more influential than Age.

4. Why this is not a single fixed “+X% purchase” like linear regression

Because logistic regression is non-linear, the change in probability (e.g., from 20% → 35% vs 70% → 85%) depends on the starting probability.

If you want to show impact in percentage points, you usually:

Pick a typical customer profile (Age₀, Income₀),

Compute baseline probability
𝑝
0
p
0
	​

,

Change one feature by +1 unit,

Recompute probability
𝑝
1
p
1
	​

,

Report Δp = p₁ − p₀ as “effect” at that profile.

In practice (for business decks), people often show:

Odds ratios table (like above), plus

1–2 example scenarios: “For a typical 30-year-old with X income, raising income by one unit increases predicted purchase probability from 25% to 60%,” etc.

If you share how you scaled Income (raw AED or standardized), I can give you a concrete “from X% to Y%” example for your model.